hoot-core-test/src/test/cpp/hoot/core/schema/ImplicitTagRulesDatabaseDeriverTest.cpp

```diff
@@ -0,0 +1,266 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2017 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+// Hoot
+#include "../TestUtils.h"
+#include <hoot/core/schema/ImplicitTagRulesDatabaseDeriver.h>
+#include <hoot/core/io/ImplicitTagRulesSqliteReader.h>
+
+// Qt
+#include <QDir>
+
+namespace hoot
+{
+
+class ImplicitTagRulesDatabaseDeriverTest : public CppUnit::TestFixture
+{
+  CPPUNIT_TEST_SUITE(ImplicitTagRulesDatabaseDeriverTest);
+  CPPUNIT_TEST(runBasicTest);
+  CPPUNIT_TEST(runMinTagOccurrencePerWordTest);
+  CPPUNIT_TEST(runMinWordLengthTest);
+  CPPUNIT_TEST(runTagIgnoreTest);
+  CPPUNIT_TEST(runWordIgnoreTest);
+  CPPUNIT_TEST(runBadInputsTest);
+  CPPUNIT_TEST(runCustomRuleTest);
+  CPPUNIT_TEST(runSchemaValuesOnlyOffTest);
+  CPPUNIT_TEST_SUITE_END();
+
+public:
+
+  static QString inDir() { return "test-files/schema/ImplicitTagRulesDatabaseDeriverTest"; }
+  static QString outDir() { return "test-output/schema/ImplicitTagRulesDatabaseDeriverTest"; }
+
+  void runBasicTest()
+  {
+    QDir().mkpath(outDir());
+
+    const QString input = inDir() + "/ImplicitTagRulesDatabaseDeriverTest-input.implicitTagRules";
+    const QString dbOutputFile =
+      outDir() + "/ImplicitTagRulesDatabaseDeriverTest-runBasicTest-out.sqlite";
+
+    ImplicitTagRulesDatabaseDeriver rulesDeriver;
+    rulesDeriver.setMinTagOccurrencesPerWord(1);
+    rulesDeriver.setMinWordLength(1);
+    rulesDeriver.setUseSchemaTagValuesForWordsOnly(true);
+    rulesDeriver.setCustomRuleFile("");
+    rulesDeriver.setTagIgnoreFile("");
+    rulesDeriver.setWordIgnoreFile("");
+    rulesDeriver.deriveRulesDatabase(input, dbOutputFile);
+
+    ImplicitTagRulesSqliteReader dbReader;
+    dbReader.open(dbOutputFile);
+    CPPUNIT_ASSERT_EQUAL(86L, dbReader.getRuleCount());
+    dbReader.close();
+  }
+
+  void runBadInputsTest()
+  {
+    QDir().mkpath(outDir());
+
+    ImplicitTagRulesDatabaseDeriver rulesDeriver;
+    rulesDeriver.setMinTagOccurrencesPerWord(1);
+    rulesDeriver.setMinWordLength(1);
+    rulesDeriver.setUseSchemaTagValuesForWordsOnly(true);
+    rulesDeriver.setCustomRuleFile("");
+    rulesDeriver.setTagIgnoreFile("");
+    rulesDeriver.setWordIgnoreFile("");
+
+    QString exceptionMsg("");
+
+    try
+    {
+      rulesDeriver.deriveRulesDatabase(
+        inDir() + "/ImplicitTagRulesDatabaseDeriverTest-input.txt",
+        outDir() + "/ImplicitTagRulesDatabaseDeriverTest-runBasicTest-out.sqlite");
```

Comment:
Using the `runBasicTest` name for the output sqlite file could cause issues when run in parallel.

Reply:
i'll change that

Reply:
The exception actually gets thrown before any file gets written, but in the case the exception doesn't get thrown and shouldn't its a good idea to keep the file names unique....so changed them.